strangelove-ventures/packet-forward-middleware
# refactor: make the middleware compliant with the ics30 spec + cleanup

The middleware was not technically meeting the criteria set out in the [spec](https://github.com/cosmos/ibc/tree/main/spec/app/ics-030-middleware) which made it impossible to integrate into the swap middleware unless we deviated from the spec there also.

I think this approach is better since anyone implementing an IBC middleware according to the spec should be able to now use the forward middleware when building their middleware stack.

Since this required touching most of the code I went ahead and restructured things to mirror the project structure of the swap middleware and the fee middleware which I think adds some nice consistency across repos and makes things easier to navigate.

Mocks and tests have been adjusted accordingly but I would like to test this in `ibctest` before merging. After we merge we will want to bring these changes into the other branches as well.

Edit: `ibctest` case for the forward middleware is passing with these changes 🎉  

To test this:

You will need to setup heighliner and clone down this branch of our Gaia fork
https://github.com/strangelove-ventures/gaia/tree/justin/forward_middleware_v3_refactor

Once that is done you can `heighliner build -c gaia --local` to build the local Docker image for Gaia

With the image built you should be able to run the updated test case here 
https://github.com/strangelove-ventures/ibctest/tree/justin/forward-middleware-refactor
